client/src/app/users/user.service.spec.ts

```diff
@@ -59,135 +60,249 @@ describe('User service: ', () => {
+            .toHaveBeenCalledTimes(1);
+          // The mocked method should have been called with two arguments:
+          //   * the appropriate URL ('/api/users' defined in the `UserService`)
+          //   * An options object containing an `HttpParams` with the `role`:`admin`
+          //     key-value pair.
+          expect(mockedMethod)
+            .withContext('talks to the correct endpoint')
+            .toHaveBeenCalledWith(userService.userUrl, { params: new HttpParams().set('role', 'admin') });
+        });
+    });
 
-    // Check that the role, company, and age parameters are correct
-    expect(req.request.params.get('role')).toEqual('editor');
-    expect(req.request.params.get('company')).toEqual('IBM');
-    expect(req.request.params.get('age')).toEqual('37');
+    it('correctly calls api/users with filter parameter \'age\'', () => {
+      const mockedMethod = spyOn(httpClient, 'get').and.returnValue(of(testUsers));
+
+      userService.getUsers({ age: 25 }).subscribe((users: User[]) => {
+        expect(users)
+          .withContext('expected users')
+          .toEqual(testUsers);
+        expect(mockedMethod)
+          .withContext('one call')
+          .toHaveBeenCalledTimes(1);
+        expect(mockedMethod)
+          .withContext('talks to the correct endpoint')
+          .toHaveBeenCalledWith(userService.userUrl, { params: new HttpParams().set('age', '25') });
+      });
+    });
 
-    req.flush(testUsers);
+    it('correctly calls api/users with multiple filter parameters', () => {
+        const mockedMethod = spyOn(httpClient, 'get').and.returnValue(of(testUsers));
+
+        userService.getUsers({ role: 'editor', company: 'IBM', age: 37 }).subscribe((users: User[]) => {
+          // This test checks that the call to `userService.getUsers()` does several things:
+          //   * It returns the expected array of users (namely `testUsers` as discussed above).
+          //   * It calls the mocked method (`HttpClient#get()`) exactly once.
+          //   * It calls it with the correct endpoint (`userService.userUrl`).
+          //   * It calls it with the correct parameters:
+          //      * There should be three parameters (this makes sure that there aren't extras).
+          //      * There should be a "role:editor" key-value pair.
+          //      * And a "company:IBM" pair.
+          //      * And a "age:37" pair.
+
+          // This gets the arguments for the first (and in this case only) call to the `mockMethod`.
+          const [url, options] = mockedMethod.calls.argsFor(0);
+          // Gets the `HttpParams` from the options part of the call.
+          // `options.param` can return any of a broad number of types;
+          // it is in fact an instance of `HttpParams`, and I need to use
+          // that fact, so I'm casting it (the `as HttpParams` bit).
+          const calledHttpParams: HttpParams = (options.params) as HttpParams;
+          expect(users)
```

Comment:
And drop this?

Reply:
Here I am less confident... does this line and it's related `withContext` and `toEqual` go away? I think it does since this is a call to `getUsers` and we get all the users back in this test. Is that correct?

Reply:
Again, it's the `expect` that I'm suggesting we might drop.